src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/HybridHtmlFragmentCache.cs

```diff
@@ -0,0 +1,219 @@
+    /// services using the <code>options-distributed</code> attribute.
+    /// </summary>
+    public class HybridHtmlFragmentCache : IHtmlFragmentCache
+    {
+        private const string PriorityOptionName = "priority";
+        private const string DistributedOptionName = "distributed";
+
+        private readonly IMemoryCache _memoryCache;
+        private readonly IServiceProvider _serviceProvider;
+
+        public HybridHtmlFragmentCache(IMemoryCache memoryCache, IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+            _memoryCache = memoryCache;
+        }
+
+        public Task<IHtmlContent> SetAsync(string key, Func<Task<IHtmlContent>> renderContent, HtmlFragmentCacheContext context)
+        {
+            IDistributedCache distributedCache;
+
+            // is the distributed option set?
+            if (EnsureDistributedService(context, out distributedCache))
+            {
+                return SetDistributedAsync(key, renderContent, context, distributedCache);
+            }
+            else
+            {
+                return SetMemoryAsync(key, renderContent, context);
+            }
+        }
+
+        private async Task<IHtmlContent> SetMemoryAsync(string key, Func<Task<IHtmlContent>> renderContent, HtmlFragmentCacheContext context)
+        {
+            var options = new MemoryCacheEntryOptions();
+
+            if (context.ExpiresOn != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresOn.Value);
+            }
+
+            if (context.ExpiresAfter != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresAfter.Value);
+            }
+
+            if (context.ExpiresSliding != null)
+            {
+                options.SetSlidingExpiration(context.ExpiresSliding.Value);
+            }
+
+            if (context.HasOptions && context.Options.ContainsKey(PriorityOptionName))
+            {
+                CacheItemPriority priority;
+                if (Enum.TryParse(context.Options[PriorityOptionName], out priority))
+                {
+                    options.SetPriority(priority);
+                }
+            }
+
+            // Create an entry link scope and flow it so that any tokens related to the cache entries
+            // created within this scope get copied to this scope.
+            using (var link = _memoryCache.CreateLinkingScope())
+            {
+                var content = await renderContent();
+
+                options.AddEntryLink(link);
+
+                var stringBuilder = new StringBuilder();
+                using (var writer = new StringWriter(stringBuilder))
+                {
+                    content.WriteTo(writer, context.HtmlEncoder);
+                }
+
+                var result = new StringBuilderHtmlContent(stringBuilder);
+
+                return _memoryCache.Set(key, result, options) as IHtmlContent;
```

Comment:
`_memoryCache.Set<IHtmlContent>(key, result, options);`. This API is kinda awful to use.